nabs/simulators.py

```diff
@@ -0,0 +1,260 @@
+            if msg.obj not in staged_devices:
+                raise ValidError("Plan attempts to unstage a device "
+                                 "that has not been staged. "
+                                 f"(msg: {i})")
+            # Currently assumes devices are unstaged in reverse order
+            # of how they were staged.
+            staged_devices.pop()
+
+    # at end of plan, nothing should be left
+    if open_stack:
+        raise ValidError("Plan ended without all runs being closed.")
+    if run_keys:
+        raise ValidError("Plan ended with unmatched run keys.")
+    if staged_devices:
+        raise ValidError(
+            "Plan ended without unstaging all staged devices."
+        )
+
+
+def check_stray_calls(
+    plan: Iterator[Any],
+    patches: List[str] = default_patches
+) -> None:
+    """
+    Validate that plan does not invoke any caput functionality
+    outside of messages.
+
+    Runs the check within a multiprocessing.Process to isolate
+    namespace from the parent process.
+
+    Relies on the pre-existing knowledge of which methods make calls
+    to pyepics/caput functionality.
+
+    Parameters
+    ----------
+    plan : iterable or generator
+        Must yield `Msg` objects
+
+    Raises
+    ------
+    ValidError
+        If attempts to access any forbidden methods
+    """
+    def inner_func(plan):
+        with patch_sys_modules(patches):
+            for _ in plan:
+                continue
+
+    # run check inside multiprocess to avoid tampering with parent
+    # namespace.  Effectively an attempt at containerization.
+    p = Process(target=inner_func, args=(plan,))
+
+    # Disallow this subprocess from spawning (multiprocess) children
+    p.daemon = True
+    p.start()
+    p.join_and_raise()
+
+
+# check_limits is not hinted, so hinting this becomes miserable
+validators = [
+    check_stray_calls,
+    check_open_close,
+    check_limits,
+]
+
+
+def validate_plan(
+    plan: Generator,
+    validators=validators
+) -> Tuple[bool, str]:
+    """
+    Validate plan with all available checkers.
+
+    Parameters
+    ----------
+    plan: generator function
+        Once called, must yield `Msg` objects.
+
+    validators: list of check functions
+        functions to run on the provided plan.  These should take an
+        evaluated plan (generator) as input, and raise exceptions on
+        failure.  (should NOT take generator functions)
+
+    Returns
+    -------
+    boolean
+        Indicates if validation was successful (``True``) or failed
+        (``False``).
+    str
+        Error message that explains the reason for validation
+        failure. Empty string if validation is successful.
+
+    """
+    success, msg = True, ""
+    try:
+        plan_list = itertools.tee(plan, len(validators))
+        for plan, check in zip(plan_list, validators):
+            print(f'running {check.__name__}')
+            check(plan)
+    except Exception as ex:
+        print(ex)
```

Comment:
Intentional prints here? What does the validation output actually look like with the current proposed code?